### PR TITLE
Consensus threading and queues rebase

### PR DESCRIPTION
**Pull request by jyellick**
This is a rebasing of the PR from #1279, please see there for the relevant description.

Currently executing local tests, but pushing PR in anticipation of their passing as it was a relatively merge conflict resolution.

Signed-off-by: Jason Yellick jyellick@us.ibm.com

### CODE REVIEW

**Comment by srderson** [comment]
This PR causes an error on startup for those with an existing DB. This is because a new RockDB column family is required. Ideally we'd create the CF on the fly to avoid all existing users seeing this error and not knowing the problem. @manish-sethi - any thoughts/concerns on creating new CFs on the fly if they're missing?

```
vagrant@hyperledger-devenv:v0.0.9-19a1703:/opt/gopath/src/github.com/hyperledger/fabric/peer$ ./peer peer
21:24:23.116 [crypto] main -> INFO 001 Log level recognized 'info', set to INFO
21:24:23.118 [logging] LoggingInit -> DEBU 002 Setting default logging level to DEBUG for command 'peer'
21:24:23.120 [sysccapi] RegisterSysCC -> DEBU 003 system chaincode github.com/hyperledger/fabric/core/system_chaincode/sample_syscc registered
21:24:23.123 [eventhub_producer] AddEventType -> DEBU 004 registering block
21:24:23.123 [eventhub_producer] AddEventType -> DEBU 005 registering register
21:24:23.124 [main] serve -> INFO 007 Security enabled status: false
21:24:23.124 [main] serve -> INFO 008 Privacy enabled status: false
21:24:23.124 [eventhub_producer] start -> INFO 006 event processor started
21:24:23.125 [chaincode] NewChaincodeSupport -> INFO 009 Chaincode support using peerAddress: 0.0.0.0:30303
21:24:23.125 [main] serve -> DEBU 00a Running as validating peer - making genesis block if needed
21:24:23.126 [db] createDBIfDBPathEmpty -> DEBU 00b Is db path [/var/hyperledger/production/db] empty [false]
Error opening DB Invalid argument: Column family not found: : persistCF
panic: Could not open openchain db error = [Invalid argument: Column family not found: : persistCF]

goroutine 1 [running]:
panic(0xbc7d40, 0xc820268a40)
    /opt/go/src/runtime/panic.go:464 +0x3e6
github.com/hyperledger/fabric/core/db.GetDBHandle(0x49a001)
    /opt/gopath/src/github.com/hyperledger/fabric/core/db/db.go:115 +0x2cf
github.com/hyperledger/fabric/core/ledger.fetchBlockchainSizeFromDB(0xc820000180, 0x0, 0x0)
    /opt/gopath/src/github.com/hyperledger/fabric/core/ledger/blockchain.go:314 +0x28
github.com/hyperledger/fabric/core/ledger.newBlockchain(0x8, 0x0, 0x0)
    /opt/gopath/src/github.com/hyperledger/fabric/core/ledger/blockchain.go:52 +0x2e
github.com/hyperledger/fabric/core/ledger.newLedger(0xc8201f5388, 0x0, 0x0)
    /opt/gopath/src/github.com/hyperledger/fabric/core/ledger/ledger.go:71 +0x28
github.com/hyperledger/fabric/core/ledger.GetLedger.func1()
    /opt/gopath/src/github.com/hyperledger/fabric/core/ledger/ledger.go:65 +0x1c
sync.(*Once).Do(0x151e668, 0xfe3ad0)
    /opt/go/src/sync/once.go:44 +0xe4
github.com/hyperledger/fabric/core/ledger.GetLedger(0xc8201f5418, 0x0, 0x0)
    /opt/gopath/src/github.com/hyperledger/fabric/core/ledger/ledger.go:66 +0x3b
github.com/hyperledger/fabric/core/ledger/genesis.MakeGenesis.func1()
    /opt/gopath/src/github.com/hyperledger/fabric/core/ledger/genesis/genesis.go:43 +0x37
sync.(*Once).Do(0x151e688, 0xfe3ae0)
    /opt/go/src/sync/once.go:44 +0xe4
github.com/hyperledger/fabric/core/ledger/genesis.MakeGenesis(0x0, 0x0)
    /opt/gopath/src/github.com/hyperledger/fabric/core/ledger/genesis/genesis.go:171 +0x3b
main.serve(0x151e5c0, 0x0, 0x0, 0x0, 0x0)
    /opt/gopath/src/github.com/hyperledger/fabric/peer/main.go:455 +0xd2a
main.glob.func3(0x14e0f80, 0x151e5c0, 0x0, 0x0, 0x0, 0x0)
    /opt/gopath/src/github.com/hyperledger/fabric/peer/main.go:89 +0x41
github.com/hyperledger/fabric/vendor/github.com/spf13/cobra.(*Command).execute(0x14e0f80, 0x151e5c0, 0x0, 0x0, 0x0, 0x0)
    /opt/gopath/src/github.com/hyperledger/fabric/vendor/github.com/spf13/cobra/command.go:497 +0x62c
github.com/hyperledger/fabric/vendor/github.com/spf13/cobra.(*Command).Execute(0x14e0dc0, 0x0, 0x0)
    /opt/gopath/src/github.com/hyperledger/fabric/vendor/github.com/spf13/cobra/command.go:584 +0x46a
main.main()
    /opt/gopath/src/github.com/hyperledger/fabric/peer/main.go:307 +0x17d5
```

**Comment by manish-sethi** [comment]
@srderson - I do not see any issue in creating a column-family, if missing. I think that in openDB() method, in line opts.SetCreateIfMissing(false), changing the flag to 'true' should be enough. 
If we go this way then it would be better that we cleanup CreateDB() method (for removing the code that creates the column-families and closes the db) and merge the code from openDB().

@jyellick, @corecode - just a cosmetic change if it is possible to name 'persistCF' as something more meaningful. As of now, the name of the column-families are based on what is being persisted.
Also, from performance point of view, it would be good to know about the write pattern. Approximately what is the size and frequency of a single commit to this column-family. (I am assuming that the writes to this column family are independent). I want to do some experiments with rocks db some time in future and this would be good to know from that point of view. - Thanks

File: consensus/obcpbft/obc-sieve.go [hunk 1]
```diff
@@ -714,6 +718,7 @@ func (op *obcSieve) commit() {
 func (op *obcSieve) restoreBlockNumber() {
 	var err error
 	op.blockNumber, err = op.stack.GetBlockchainSize()
+	op.blockNumber-- // The highest block number is one less than the size
```
Comment (corecode) [inline consensus/obcpbft/obc-sieve.go:721]: Is this correct?  I changed sieve to just carry the blockchainsize as blockNumber.

Comment (corecode) [inline consensus/obcpbft/obc-sieve.go:721]: Nevermind - it doesn't matter.

**Comment by corecode** [comment]
@manish-sethi The idea was to use this as a generic persistency store for per-node state which is not located in the blockchain.  Currently only consensus uses it, but I could imagine other parts of the peer using it as well.  I wanted to avoid setting precedent that would lead to many CF being created.

**Comment by jyellick** [comment]
@manish-sethi If we specify `opts.SetCreateIfMissing(true)` in `openDB`, then it seems like we could simply remove all of the DB creation code entirely?

**Comment by manish-sethi** [comment]
@jyellick - yes, you are right the code for db creation can be removed. Though, I do not recall whether the 'dir' should have been already created or rocksdb would create whole path.

**Comment by jyellick** [comment]
@manish-sethi After a little further investigation, what was really required was `opts.SetCreateIfMissingColumnFamilies(true)`  Because this change does not automatically create the DB, I decided to go ahead and leave the existing `CreateDB` logic in place.  Please review dad6b2a if you get the chance.

**Comment by manish-sethi** [comment]
@jyellick - yes, the changes look good to me. Thanks.

**Comment by tuand27613** [comment]
LGTM as well. thanks @jyellick

**Comment by srderson** [comment]
This PR has been reviewed by @jyellick @corecode @bcbrock @manish-sethi @srderson @binhn and @tuand27613. An explanation of the PR can be found in PR #1279. LGTM apart from the size which resulted in review delays. PRs of this size are to be avoided in the future.

**Comment by christo4ferris** [comment]
I'd go as far as to say must be avoided in the future

**Comment by corecode** [comment]
Can you suggest a viable development process that will allow refactoring and bug fixes that build on top of each other?  We could just submit small PRs and then wait until they get merged before submitting another?

**Comment by jyellick** [comment]
@christo4ferris @srderson Completely agree we need smaller PRs, when they are this size review becomes unmanagable.

Hoping that the CI system will mitigate this to some extent, as hopefully PRs won't sit in limbo for so long, (requiring new PRs to derive from them, and the inevitable rebasing cascade on conflict), but I'm also curious as to your response to @corecode's question, for those changesets which must invariably break things before they get better, what is the preferred method of approaching them?

**Comment by christo4ferris** [comment]
CI can help as we add more and more tests, to increase confidence level, but no one wants to review PRs that involve 4k LOC deltas - they take too long to vet which further compounds the problem to which @corecode alludes. Smaller PRs can be reviewed and merged more quickly (especially as we get more people reviewing), and rebasing is a fact of life in open source.

**Comment by jyellick** [comment]
Sorry if I was unclear, I didn't mean to imply that CI would help with large changesets, rather, that it would help getting smaller changesets in faster.  (And therefore reduce the pain of submitting many small PRs).

**Comment by christo4ferris** [comment]
@jyellick then we are on the same page... and that definitely addresses @corecode concern, IMO.

**Pull request merged**